src/Conjure/UI/Model.hs

```diff
@@ -1848,6 +1836,103 @@ rule_DotLtLeq = "generic-DotLtLeq" `namedRule` theRule where
+rule_Flatten_Lex = "flatten-lex" `namedRule` theRule where
+    theRule [essence| &a <lex &b |] = do
+      ta <- typeOf a
+      tb <- typeOf b
+      case (ta, tb) of
+        (TypeList TypeInt{}, TypeList TypeInt{}) ->
+          na "rule_Flatten_Lex" 
+        (TypeMatrix TypeInt{} TypeInt{}, TypeMatrix TypeInt{} TypeInt{}) ->
+          na "rule_Flatten_Lex"
+        _ -> return () 
+      fa <- flatten a
+      fb <- flatten b
+      tfa <- typeOf fa
+      tfb <- typeOf fb
+      case (tfa, tfb) of
+        (TypeList TypeInt{}, TypeList TypeInt{}) -> return ()
+        (TypeMatrix TypeInt{} TypeInt{}, TypeMatrix TypeInt{} TypeInt{}) -> return ()
+        _ -> bug $ "flattener: " <+> vcat [stringToDoc $ show tfa, stringToDoc $ show tfb]
+      return ( "Flatten Lex less"
+             , return [essence| &fa <lex &fb |]
+             )
+    theRule [essence| &a <=lex &b |] = do
+      ta <- typeOf a
+      tb <- typeOf b
+      case (ta, tb) of
+        (TypeList TypeInt{}, TypeList TypeInt{}) ->
+          na "rule_Flatten_Lex" 
+        (TypeMatrix TypeInt{} TypeInt{}, TypeMatrix TypeInt{} TypeInt{}) ->
+          na "rule_Flatten_Lex"
+        _ -> return () 
+      fa <- flatten a
+      fb <- flatten b
+      tfa <- typeOf fa
+      tfb <- typeOf fb
+      case (tfa, tfb) of
+        (TypeList TypeInt{}, TypeList TypeInt{}) -> return ()
+        (TypeMatrix TypeInt{} TypeInt{}, TypeMatrix TypeInt{} TypeInt{}) -> return ()
+        _ -> bug $ "flattener: " <+> vcat [stringToDoc $ show tfa, stringToDoc $ show tfb]
+      return ( "Flatten Lex Lt"
+             , return [essence| &fa <=lex &fb |]
+             )
+    theRule _ = na "rule_Flatten_Lex"  
+    flatten a = do
+      ta <- typeOf a
+      case ta of
+        TypeBool -> return [essence| [-toInt(&a)] |]
+        TypeInt{} -> return [essence| [&a] |]
+        TypeList TypeInt{} -> return a
+        TypeMatrix TypeInt{} TypeInt{} -> return a
+        _ ->
+          case a of
+            AbstractLiteral x -> do
+              case x of
+                AbsLitTuple xs -> do
+                  fxs <- sequence (flatten <$> xs)
+                  let flatxs = fromList fxs
+                  return [essence| flatten(&flatxs) |]
+                AbsLitMatrix _ xs -> do
+                  fxs <- sequence (flatten <$> xs)
+                  let flatxs = fromList fxs
+                  return [essence| flatten(&flatxs) |]
+                _ -> bug $ "epilogue: flattenLex: isn't defined for this abslit fellow..."
+                    <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+            Constant c ->
+              case c of
+                ConstantAbstract ca ->
+                  case ca of
+                    AbsLitTuple xs -> do
+                      fxs <- sequence (flatten <$> (Constant <$> xs))
+                      let flatxs = fromList fxs
+                      return [essence| flatten(&flatxs) |]
+                    AbsLitMatrix _ [] -> return [essence| [0] |]
```

Comment:
I have changed this to rewrite to ([] : `matrix indexed by [int()] of int`) - waiting on tests to finish before changing the output files back.